File: EvolutionaryGenerator_v2.1_Java/Generator/src/mx/tec/hermes/problems/kp/Knapsack.java
package mx.tec.hermes.problems.kp;

import java.util.LinkedList;
import java.util.List;

/**
 * Provides the methods to create and use knapsacks for the knapsack problem.
 * 
 * @author José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
 * @version 2.0
 */
public class Knapsack {

    private int capacity;
    private double sProfit, pProfit;
    private final List<Item> items;

    /**
     * Creates a new instance of <code>Knapsack</code>.
     *
     * @param capacity The capacity of this knapsack.
     */
    public Knapsack(int capacity) {
        this.capacity = capacity;
        sProfit = 0;
        pProfit = 1;
        items = new LinkedList();
    }    
    
    /**
     * Returns the current capacity of this knapsack.
     *
     * @return The current capacity of this knapsack.
     */
    public int getCapacity() {
        return capacity;
    }   
    
    /**
     * Returns the current sum of profits in this knapsack.
     *
     * @return The current sum of profits in this knapsack.
     */
    public double getSumOfProfit() {
        return sProfit;
    }
    
    /**
     * Returns the current product of profits in this knapsack.
     *
     * @return The current product of profits in this knapsack.
     */
    public double getProductOfProfit() {
        return pProfit;
    }
   
    int[] getSolution(int nbItems) {
        int[] solution = new int[nbItems];
        for (Item item : items) {
            solution[item.getId()] = 1;
        }
        return solution;
    }
    
    /**
     * Revises if the item provided can be packed in this knapsack.
     *
     * @param item The item to be packed.
     * @return <code>true</code> if the item can be packed in this knapsack, <code>false</code> otherwise.
     */
    public boolean canPack(Item item) {
        return item.getWeight() <= getCapacity();
    }

    /**
     * Packs an item into this knapsack.
     *
     * @param item The item to pack.
     * @return <code>true</code> if the item was successfully packed, <code>false</code> otherwise.
     */
    public boolean pack(Item item) {
        if (item.getWeight() <= capacity) {
            items.add(item);
            capacity -= item.getWeight();
            sProfit += item.getProfit();
            pProfit *= item.getProfit();
            return true;
        }
        return false;
    }

    /**
     * Clones this knapsack.
     * 
     * @return A deep copy of this knapsack-
     */
    public Knapsack copy() {
        Knapsack tmp;
        tmp = new Knapsack(capacity);
        for (Item item : items) {
            tmp.pack(item);
        }
        return tmp;
    }
    
    /**
     * Returns the string representation of this knapsack.
     *
     * @return The string representation of this knapsack.
     */
    @Override
    public String toString() {
        StringBuilder string;
        string = new StringBuilder();
        for (Item item : items) {
            string.append(item.toString()).append(" ");
        }
        return string.toString().trim();
    }

}

File: EvolutionaryGenerator_v2.1_Python/hermes/problems/kp/knapsack.py
"""Translation of ``mx.tec.hermes.problems.kp.Knapsack``."""

from hermes.problems.kp.item import Item


class Knapsack:
    """Provides the methods to create and use knapsacks for the knapsack problem.

    :author: Paola Azeneth Castillo Gutiérrez (Python translation)
    :original author: José Carlos Ortiz Bayliss (jcobayliss@tec.mx)
    :original version: 2.0
    """

    def __init__(self, capacity: int):
        """Creates a new instance of ``Knapsack``.

        :param capacity: The capacity of this knapsack.
        """
        self.capacity = capacity
        self.s_profit = 0.0
        self.p_profit = 1.0
        self.items = []

    def get_capacity(self) -> int:
        """Returns the current capacity of this knapsack.

        :return: The current capacity of this knapsack.
        """
        return self.capacity

    def get_sum_of_profit(self) -> float:
        """Returns the current sum of profits in this knapsack.

        :return: The current sum of profits in this knapsack.
        """
        return self.s_profit

    def get_product_of_profit(self) -> float:
        """Returns the current product of profits in this knapsack.

        :return: The current product of profits in this knapsack.
        """
        return self.p_profit

    def get_solution(self, nb_items: int) -> list:
        """Returns the current solution as a vector of zeros and ones.

        :param nb_items: The number of items in the problem this knapsack belongs to.
        :return: The current solution to the problem this knapsack belongs to.
        """
        solution = [0] * nb_items
        for item in self.items:
            solution[item.get_id()] = 1
        return solution

    def can_pack(self, item: Item) -> bool:
        """Revises if the item provided can be packed in this knapsack.

        :param item: The item to be packed.
        :return: ``True`` if the item can be packed in this knapsack, ``False`` otherwise.
        """
        return item.get_weight() <= self.get_capacity()

    def pack(self, item: Item) -> bool:
        """Packs an item into this knapsack.

        :param item: The item to pack.
        :return: ``True`` if the item was successfully packed, ``False`` otherwise.
        """
        if item.get_weight() <= self.capacity:
            self.items.append(item)
            self.capacity -= item.get_weight()
            self.s_profit += item.get_profit()
            self.p_profit *= item.get_profit()
            return True
        return False

    def copy(self) -> "Knapsack":
        """Clones this knapsack.

        :return: A deep copy of this knapsack.
        """
        tmp = Knapsack(self.capacity)
        for item in self.items:
            tmp.pack(item)
        return tmp

    def __str__(self) -> str:
        """Returns the string representation of this knapsack.

        :return: The string representation of this knapsack.
        """
        string = []
        for item in self.items:
            string.append(str(item) + " ")
        return "".join(string).strip()
